lib/cc/resolv_override.rb

```diff
@@ -0,0 +1,46 @@
+require "resolv-replace"
+
+module CC
+  # Allows for control of how DNS is resolved so that when we resolve a host for
+  # the purposes of checking that it's not internal, we can ensure that future
+  # resolutions (done as part of making the request) will find the same IP
+  # address.
+  module ResolvOverride
+    def self.get_and_fix_address(host)
+      dns.getaddress(host).tap do |address|
+        fixed.add_address(host, address)
+      end
+    end
+
+    def self.dns
+      @dns ||= Resolv::DNS.new
+    end
+
+    def self.fixed
+      @fixed ||= ResolvOverride::FixedAddress.new
+    end
+
+    class FixedAddress < Resolv::DNS
+      def add_address(host, address)
+        addresses[host] = address
+      end
+
+      def each_address(name, &proc)
+        if addresses.key?(name)
+          yield addresses.fetch(name)
+        end
+      end
+
+      private
+
+      def addresses
+        @addresses ||= {}
+      end
+    end
+  end
+end
+
+Resolv::DefaultResolver.replace_resolvers([
+  CC::ResolvOverride.fixed,
+  CC::ResolvOverride.dns,
```

Comment:
At present, doing nothing else, the resolv-override is going to be active in any app container (since it loads the cc-services code, which loads this code). So if we dropped `dns` here, other calls (like App hitting GH API or something) would also go by way of `fixed` which (currently) doesn't fall back if you haven't pre-set the host->address case.

So that would all have to be addressed somehow; then there might be other interactions I'm _not_ thinking about. Ultimately, I'm not sure the of the upside.